### DISCUSSION TITLE
Insufficient Authentication

### DISCUSSION THREAD

**Discussion by jasonrepos** [hbldh/bleak#660] [Q&A]
* bleak version: 0.11.0
* Python version: Python 3.7.3
* Operating System: Raspbian 10
* BlueZ version (`bluetoothctl -v`) in case of Linux: bluetoothctl: 5.50

### Description

I am trying to write a python script to control my HEATSTORE heater, which I will eventually try to get into homeassistant through bt-mqtt-gateway.
The HEATSTORE heater does have a pairing function which then requires the input of a predetermined pin which consists of 6 digits.

### What I Did

I have paired the heater with Linux using bluetoothctl and used gatttool to grab a value from a handle with no issue.

```
pi@raspberrypi:~/bt-mqtt-gateway/workers/bleak $ gatttool -i hci0 --sec-level=medium -b 00:A0:50:4E:23:8B -I
[00:A0:50:4E:23:8B][LE]> connect
Attempting to connect to 00:A0:50:4E:23:8B
Connection successful
[00:A0:50:4E:23:8B][LE]> primary
attr handle: 0x0001, end grp handle: 0x0009 uuid: 00001800-0000-1000-8000-00805f9b34fb
attr handle: 0x000a, end grp handle: 0x000d uuid: 00001801-0000-1000-8000-00805f9b34fb
attr handle: 0x000e, end grp handle: 0x0011 uuid: 00060000-f8ce-11e4-abf4-0002a5d5c51b
attr handle: 0x0012, end grp handle: 0x0015 uuid: 00060000-f8ce-11e4-abf4-0002a5d5c51c
attr handle: 0x0016, end grp handle: 0x003e uuid: 00000000-0000-1000-8000-00805f9b34fb
attr handle: 0x0041, end grp handle: 0x004f uuid: 00000000-0001-1000-8000-00805f9b34fb
attr handle: 0x0050, end grp handle: 0x0062 uuid: 00000000-0002-1000-8000-00805f9b34fb
attr handle: 0x0063, end grp handle: 0x0089 uuid: 00000000-0003-1000-8000-00805f9b34fb
[00:A0:50:4E:23:8B][LE]> char-read-hnd 0x0009
Characteristic value/descriptor: 00 
[00:A0:50:4E:23:8B][LE]> 
```

```
[bluetooth]# info 00:A0:50:4E:23:8B
Device 00:A0:50:4E:23:8B (public)
        Name: AL2381<Heatstore>
        Alias: AL2381<Heatstore>
        Paired: yes
        Trusted: yes
        Blocked: no
        Connected: no
        LegacyPairing: no
        UUID: Unknown                   (00000000-0000-1000-8000-00805f9b34fb)
        UUID: Vendor specific           (00000000-0001-1000-8000-00805f9b34fb)
        UUID: Vendor specific           (00000000-0002-1000-8000-00805f9b34fb)
        UUID: Vendor specific           (00000000-0003-1000-8000-00805f9b34fb)
        UUID: Generic Access Profile    (00001800-0000-1000-8000-00805f9b34fb)
        UUID: Generic Attribute Profile (00001801-0000-1000-8000-00805f9b34fb)
        UUID: Vendor specific           (00060000-f8ce-11e4-abf4-0002a5d5c51b)
        UUID: Vendor specific           (00060000-f8ce-11e4-abf4-0002a5d5c51c)
[bluetooth]#
```

I then started to try and get something together with bleak and have found in the tcpdump that I am getting "Insufficient Authentication" and what looks to be the device requesting authentication, which I can only imagine is the pairing PIN.

Here is the bleak script - heat.py:-

```
import asyncio
from bleak import BleakClient
from pprint import pprint

address = "00:A0:50:4E:23:8B"
test_handle = "0x0009"

async def run(address):
    client = BleakClient(address)
    try:
        await client.connect()
        await client.pair(3)
        services = await client.read_gatt_char(test_handle)
        pprint(services)
    except Exception as e:
        print(e)
    finally:
        await client.disconnect()

loop = asyncio.get_event_loop()
loop.run_until_complete(run(address))
```

Output from the console for the script heat.py:-

```
pi@raspberrypi:~/bt-mqtt-gateway/workers/bleak $ python3 heat.py
Characteristic with UUID 0x0009 could not be found!
pi@raspberrypi:~/bt-mqtt-gateway/workers/bleak $
```

Attached are the tcpdumps which shows the behaviour from both the bleak script and the gatttool commands.
[tcpdumps.zip](https://github.com/hbldh/bleak/files/6253256/tcpdumps.zip)

Any idea how I can get this to work?  Is bleak meant to be using the pairing from bluetootctl or am I mean to pair through Bleak?

**Comment by hbldh** [comment]
The output you got seems to work fine until you get to the `read_gatt_char` part, which means you do not need to pair in Bleak. (Given that you have already done that in `bluetoothctl`...) Bleak will see the `Paired: yes` property and just exit the `pair` method. The `3` you sent in will not be used for anything by the way; there is a protection level in Windows, but not in BlueZ.

It seems you specified your handle as a hex string, making Bleak interpret it as uuid string. If you change it to
```python
test_handle = 0x0009  # integer value 9 will be stored in test_handle instead of string "0x0009"
```
it should work.
